sc-memory/sc-memory/src/scs/scs_parser.cpp

```diff
@@ -302,39 +309,57 @@ bool Parser::Parse(std::string const & str)
   return result;
 }
 
-ParsedElement & Parser::GetParsedElementRef(ElementHandle const & elID)
+ParsedElement & Parser::GetParsedElementRef(ElementHandle const & handle)
 {
-  auto & container = (elID.IsLocal() ? m_parsedElementsLocal : m_parsedElements);
+  ParsedElementVector & container = GetContainerByElementVisibilityRef(handle.GetVisibility());
 
-  if (*elID >= container.size())
-  {
+  if (*handle >= container.size())
     SC_THROW_EXCEPTION(
         utils::ExceptionItemNotFound,
-        std::string("ElementId{") + std::to_string(*elID) + ", " + std::to_string(elID.IsLocal()) + "}");
-  }
+        "Parsed element not found. "
+        "ElementId: {"
+            << std::to_string(*handle) << ", " << (size_t)handle.GetVisibility() << "}. Valid range: [0, "
+            << container.size() - 1 << "].");
 
-  return container[*elID];
+  return container[*handle];
 }
 
-ParsedElement const & Parser::GetParsedElement(ElementHandle const & elID) const
+ParsedElement const & Parser::GetParsedElement(ElementHandle const & handle) const
 {
-  auto & container = (elID.IsLocal() ? m_parsedElementsLocal : m_parsedElements);
+  ParsedElementVector const & container = GetContainerByElementVisibility(handle.GetVisibility());
 
-  if (*elID >= container.size())
-  {
+  if (*handle >= container.size())
     SC_THROW_EXCEPTION(
         utils::ExceptionItemNotFound,
-        std::string("ElementId{") + std::to_string(*elID) + ", " + std::to_string(elID.IsLocal()) + "}");
-  }
+        "Error: Parsed element not found. "
+        "ElementId: {"
+            << std::to_string(*handle) << ", " << (size_t)handle.GetVisibility() << "}. Valid range: [0, "
+            << container.size() - 1 << "].");
 
-  return container[*elID];
+  return container[*handle];
 }
 
 Parser::TripleVector const & Parser::GetParsedTriples() const
 {
   return m_parsedTriples;
 }
 
+void Parser::ForEachGeneratableTriple(
+    std::function<void(ParsedElement const &, ParsedElement const &, ParsedElement const &)> const & callback) const
+{
+  for (auto const & triple : m_parsedTriples)
+  {
+    auto const & source = GetParsedElement(triple.m_source);
+    auto const & connector = GetParsedElement(triple.m_connector);
+    auto const & target = GetParsedElement(triple.m_target);
+
+    if (source.IsElementType() && (connector.GetType() == ScType::ConstPermPosArc))
```

Comment:
```suggestion
    if (m_elementTypeArcs.count(connector.GetIdtf()))
```
will this be faster?

Reply:
now it checks for connector.IsElementType() which always returns false

Reply:
the same method is used